presto-parquet/src/main/java/io/prestosql/parquet/reader/LongDecimalColumnReader.java

```diff
@@ -41,20 +43,32 @@
     @Override
     protected void readValue(BlockBuilder blockBuilder, Type prestoType)
     {
-        if (!isLongDecimal(prestoType)) {
+        if (!(prestoType instanceof DecimalType)) {
             throw new ParquetDecodingException(format("Unsupported Presto column type (%s) for Parquet column (%s)", prestoType, columnDescriptor));
         }
+
         DecimalType prestoDecimalType = (DecimalType) prestoType;
-        if (prestoDecimalType.getScale() != parquetDecimalType.getScale()) {
-            throw new ParquetDecodingException(format(
-                    "Presto decimal column type has different scale (%s) than Parquet decimal column (%s)",
-                    prestoDecimalType.getScale(),
-                    parquetDecimalType.getScale()));
-        }
 
         if (definitionLevel == columnDescriptor.getMaxDefinitionLevel()) {
-            Binary value = valuesReader.readBytes();
-            prestoType.writeSlice(blockBuilder, Decimals.encodeUnscaledValue(new BigInteger(value.getBytes())));
+            Binary binary = valuesReader.readBytes();
+            Slice value = Decimals.encodeUnscaledValue(new BigInteger(binary.getBytes()));
+
+            if (prestoDecimalType.isShort()) {
+                prestoType.writeLong(blockBuilder, longToShortCast(
+                        value,
+                        parquetDecimalType.getPrecision(),
+                        parquetDecimalType.getScale(),
+                        prestoDecimalType.getPrecision(),
+                        prestoDecimalType.getScale()));
+            }
+            else {
+                prestoType.writeSlice(blockBuilder, longToLongCast(
```

Comment:
In `longToLongCast`, add:
```
if (sourcePrecision == resultPrecision && sourceScale == resultScale) {
    return value;
}
```

this method doesn't cleanly shortcircuit when it's a no-op.